sources/engine/Stride.BepuPhysics/Stride.BepuPhysics.Navigation/Components/RecastNavigationComponent.cs

```diff
@@ -120,30 +117,38 @@ private void Move(float deltaTime)
             return;
         }
 
-        var position = Entity.Transform.WorldMatrix.TranslationVector;
+        var targetPosition = Entity.Transform.WorldMatrix.TranslationVector;
 
         var nextWaypointPosition = Path[0];
-        var distanceToWaypoint = Vector3.Distance(position, nextWaypointPosition);
-
-        // When the distance between the character and the next waypoint is large enough, move closer to the waypoint
-        if (distanceToWaypoint > 0.1)
+        targetPosition = Vector3.MoveTo(targetPosition, nextWaypointPosition, Speed * deltaTime);
+        if (targetPosition == nextWaypointPosition && Path.Count > 0)
         {
-            var direction = nextWaypointPosition - position;
-            direction.Normalize();
-            direction *= Speed * deltaTime;
-
-            position += direction;
+            // need to test if storing the index in Pathfinder would be faster than this.
+            Path.RemoveAt(0);
         }
-        else
+
+        // Handle the scenario where the agent has a parent.
+        ParentToLocal(ref targetPosition, Entity);
```

Comment:
There's an extension method specifically to transform from world to local, see `Stride.Engine.EntityTransformExtensions.WorldToLocal()`, right now yours doesn't take scaling or rotation into account.

Reply:
For some reason I cant get WorldToLocal to do what I want. Ill take a deeper look but the value that gets returned seems to be wrong at least with how Im trying to use it.

This entity has no parent so it should be world space like implementation has.

World To Local : 
- start: {X:39.560036 Y:15.146196 Z:12.773775}
- result: {X:0 Y:-0.038010597 Z:0}

```csharp
targetPosition = Entity.Transform.WorldToLocal(targetPosition);
Entity.Transform.Position = targetPosition;
```

Parent To Local: 
- start: {X:39.560036 Y:15.146196 Z:12.773775}
- result: {X:39.560036 Y:15.133001 Z:12.773775}

```csharp
ParentToLocal(ref targetPosition, Entity);
Entity.Transform.Position = targetPosition;
```

Reply:
Remember that `Transform.Position` is the position of the transform local to the parent. 
`Transform.WorldToLocal(x)` transforms the position to be local to the transform itself, not the parent's transform.

Reply:
Thats the weird part, if there is no parent it should be returning its own coordinates which would be offset from the world origin right? It seems to be treating it like there is a parent when there isnt. Ill see if I can recreate it in the demo when I have a chance but that method seems to be very odd at least with how I want to use it.

Reply:
> Thats the weird part, if there is no parent it should be returning its own coordinates which would be offset from the world origin right ?

I don't think my explanation was clear enough, you're not supposed to do
```cs
targetPosition = Entity.Transform.WorldToLocal(targetPosition);
Entity.Transform.Position = targetPosition;
```
What you should be doing is
```cs
targetPosition = Entity.Transform.Parent.WorldToLocal(targetPosition);
Entity.Transform.Position = targetPosition;
```
`Entity.Transform.Position` is the entity's local position inside its parent. `WorldToLocal(targetPosition)` transforms the position to be local to itself.
For `myTransform` with position `{0,1,0}` without a parent whose scene doesn't have any translation, using `myTransform.WorldToLocal({0,1,0})` will return `{0,0,0}`, if you want to set the local position of one of its child to match the world position {0,1,0} it would be a local position of zero as the parent is at that position - there is no translation to perform for the child to be there as well.

And as I alluded above, scenes also have a position offset property, entities without parent still require additional translation based on their scenes matrix